tests/Integration/Behaviour/Features/Context/Domain/CartFeatureContext.php

```diff
@@ -584,7 +584,7 @@ public function useDiscountVoucherOnCart(string $voucherCode, float $discountAmo
     public function useDiscountByCodeOnCart(string $voucherCode, string $cartReference)
     {
         $cartId = SharedStorage::getStorage()->get($cartReference);
-        $cartRuleId = SharedStorage::getStorage()->get($voucherCode);
+        $cartRuleId = (int) CartRule::getIdByCode($voucherCode);
```

Comment:
Why do we rely on DB code value now? If it worked fine ith the reference system we should continue to use it no? it's preferable to rely on a reference bound to a test scenario than a code in database that can be modified, the reference would still point to the same object while we'd need to adapt all following steps to match the new code

Reply:
Because we don't save cart rule reference by code anymore (as that was custom step), but I can set the reference everytime we create a rule when code is not empty and that will work as before then.

Reply:
Hmm then I think we can use another reference to save the ID based on the code like after addition and edition:
```php
if (!empty($cartRule->code)) {
   $this->getStorage()->set('cart_rule_code' . $cartRule->code, $cartRule->id);
}
```
then there would be to ways to get the cart rule ID:
```php
$cartRuleId = $this->getStorage()->get('cart_rule_code' . $cartRuleCode);
$cartRuleId = $this->getStorage()->get($cartRuleReference);
```
this allows you to keep the previous steps that were based on the code as reference but at least only IDs are stored in every case. And we could even add some helper method `getCartRuleIdByReference($reference): int` and `getCartRuleIdByCode($code): int`

Reply:
I just saved by the code whenever is not empty, that should be enough